src/perf/sample/writer.cpp

```diff
@@ -208,9 +208,26 @@ void Writer::end()
     if (cpuid_ == -1)
     {
         // time::now() can sometimes can be in the past :-(
-        last_time_point_ = std::max(last_time_point_, lo2s::time::now());
+        auto now = lo2s::time::now();
+        last_time_point_ = std::max(last_time_point_, now);
+
+        // If we have never written any samples on this location, we also never
+        // got to write the thread_begin event.  Make sure we do that now.
+        if (first_event_)
+        {
+            first_time_point_ = std::min(first_time_point_, now);
+            otf2_writer_ << otf2::event::thread_begin(first_time_point_, trace_.process_comm(pid_),
+                                                      -1);
+
+            // Make sure that our last timepoint is *not* before the first
+            // timepoint.  This way, events on this location span a non-negative
+            // ammount of time.
+            last_time_point_ = std::max(last_time_point_, first_time_point_);
```

Comment:
```suggestion      
```

This check is also not required. A few lines up, we make sure that `last_time_point_ >= now` and `now >= _first_time_point_`. Therefore, here `last_time_point_` is already greater (or equal) to `first_time_point_`.

Reply:
Transitive properties are hard.